File: adapters/typescript/src/logicalValue.ts
/**
 * LogicalValue grammar parser for canonical CBOR conformance test suite.
 */

/**
 * Represents a logical value in the conformance test grammar.
 * This is a recursive union type matching logical-value.schema.json.
 */
export type LogicalValue =
  | IntValue
  | FloatValue
  | TextValue
  | BytesValue
  | BoolValue
  | NullValue
  | ArrayValue
  | MapValue
  | TagValue
  | BignumValue;

interface IntValue {
  type: "int";
  value: string;
}

interface FloatValue {
  type: "float";
  width: "auto" | "f16" | "f32" | "f64";
  value: string;
}

interface TextValue {
  type: "text";
  value: string;
}

interface BytesValue {
  type: "bytes";
  value: string;
}

interface BoolValue {
  type: "bool";
  value: boolean;
}

interface NullValue {
  type: "null";
}

interface ArrayValue {
  type: "array";
  items: LogicalValue[];
}

interface MapValue {
  type: "map";
  entries: [LogicalValue, LogicalValue][];
}

interface TagValue {
  type: "tag";
  tag: bigint;
  value: LogicalValue;
}

interface BignumValue {
  type: "bignum";
  sign: "positive" | "negative";
  value: string;
}

/**
 * Parse error with explanatory message.
 */
export class ParseError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ParseError";
  }
}

/**
 * Parse a single line of logical-value JSON grammar.
 *
 * Handles the JSON number precision hazard for tag values:
 * - Rewrites "tag":<digits> to "tag":"<digits>" before parsing
 * - Converts tag string back to bigint after parsing
 * This ensures tag values >= 2^53 are preserved exactly.
 */
export function parseLogicalValueLine(line: string): LogicalValue {
  // Apply the tag-precision regex fix: rewrite "tag":<digits> to "tag":"<digits>"
  // This pattern is safe against false positives because:
  // - additionalProperties: false in tagValue schema
  // - unescaped colons can only appear as field separators
  // - escaped quotes in strings are always \"
  const fixedLine = line.replace(/"tag"\s*:\s*(\d+)/g, '"tag":"$1"');

  let json: any;
  try {
    json = JSON.parse(fixedLine);
  } catch (e) {
    throw new ParseError(`JSON parse error: ${e instanceof Error ? e.message : String(e)}`);
  }

  return parseLogicalValue(json);
}

/**
 * Parse a JSON object into a LogicalValue.
 * Assumes input is already parsed (not a string).
 */
function parseLogicalValue(json: any): LogicalValue {
  if (typeof json !== "object" || json === null || Array.isArray(json)) {
    throw new ParseError("expected a JSON object");
  }

  const type = json.type;
  if (typeof type !== "string") {
    throw new ParseError('missing or invalid "type" field');
  }

  switch (type) {
    case "int":
      return parseIntValue(json);
    case "float":
      return parseFloatValue(json);
    case "text":
      return parseText(json);
    case "bytes":
      return parseBytes(json);
    case "bool":
      return parseBool(json);
    case "null":
      return parseNull(json);
    case "array":
      return parseArray(json);
    case "map":
      return parseMap(json);
    case "tag":
      return parseTag(json);
    case "bignum":
      return parseBignum(json);
    default:
      throw new ParseError(`unknown logical-value type: ${JSON.stringify(type)}`);
  }
}

function parseIntValue(json: any): IntValue {
  const value = json.value;
  if (typeof value !== "string") {
    throw new ParseError('int: missing or invalid "value"');
  }
  return { type: "int", value };
}

function parseFloatValue(json: any): FloatValue {
  const width = json.width;
  if (!["auto", "f16", "f32", "f64"].includes(width)) {
    throw new ParseError('float: invalid or missing "width"');
  }
  const value = json.value;
  if (typeof value !== "string") {
    throw new ParseError('float: missing or invalid "value"');
  }
  return { type: "float", width: width as "auto" | "f16" | "f32" | "f64", value };
}

function parseText(json: any): TextValue {
  const value = json.value;
  if (typeof value !== "string") {
    throw new ParseError('text: missing or invalid "value"');
  }
  return { type: "text", value };
}

function parseBytes(json: any): BytesValue {
  const value = json.value;
  if (typeof value !== "string") {
    throw new ParseError('bytes: missing or invalid "value"');
  }
  return { type: "bytes", value };
}

function parseBool(json: any): BoolValue {
  const value = json.value;
  if (typeof value !== "boolean") {
    throw new ParseError('bool: missing or invalid "value"');
  }
  return { type: "bool", value };
}

function parseNull(json: any): NullValue {
  return { type: "null" };
}

function parseArray(json: any): ArrayValue {
  const items = json.items;
  if (!Array.isArray(items)) {
    throw new ParseError('array: missing or invalid "items"');
  }
  return { type: "array", items: items.map(parseLogicalValue) };
}

function parseMap(json: any): MapValue {
  const entries = json.entries;
  if (!Array.isArray(entries)) {
    throw new ParseError('map: missing or invalid "entries"');
  }
  const parsed: [LogicalValue, LogicalValue][] = [];
  for (const entry of entries) {
    if (!Array.isArray(entry) || entry.length !== 2) {
      throw new ParseError("map entry must be a 2-element array");
    }
    parsed.push([parseLogicalValue(entry[0]), parseLogicalValue(entry[1])]);
  }
  return { type: "map", entries: parsed };
}

function parseTag(json: any): TagValue {
  const tag = json.tag;
  if (typeof tag !== "bigint" && typeof tag !== "string") {
    throw new ParseError('tag: missing or invalid "tag" field');
  }
  let tagValue: bigint;
  if (typeof tag === "string") {
    try {
      tagValue = BigInt(tag);
    } catch (e) {
      throw new ParseError('tag: "tag" field is not a valid bigint string');
    }
  } else {
    tagValue = tag as bigint;
  }
  const value = json.value;
  if (!value) {
    throw new ParseError('tag: missing "value"');
  }
  return { type: "tag", tag: tagValue, value: parseLogicalValue(value) };
}

function parseBignum(json: any): BignumValue {
  const sign = json.sign;
  if (!["positive", "negative"].includes(sign)) {
    throw new ParseError('bignum: invalid or missing "sign"');
  }
  const value = json.value;
  if (typeof value !== "string") {
    throw new ParseError('bignum: missing or invalid "value"');
  }
  return { type: "bignum", sign: sign as "positive" | "negative", value };
}

File: adapters/rust/src/logical_value.rs
use serde_json::Value;

#[derive(Debug, Clone, PartialEq)]
pub enum LogicalValue {
    Int(String),
    Float { width: String, value: String },
    Text(String),
    Bytes(String),
    Bool(bool),
    Null,
    Array(Vec<LogicalValue>),
    Map(Vec<(LogicalValue, LogicalValue)>),
    Tag(u64, Box<LogicalValue>),
    Bignum { sign: String, value: String },
}

#[derive(Debug)]
pub struct ParseError(pub String);

pub fn parse(json: &Value) -> Result<LogicalValue, ParseError> {
    let obj = json.as_object().ok_or_else(|| ParseError("expected a JSON object".to_string()))?;
    let t = obj
        .get("type")
        .and_then(Value::as_str)
        .ok_or_else(|| ParseError("missing \"type\" field".to_string()))?;
    match t {
        "int" => {
            let v = obj.get("value").and_then(Value::as_str)
                .ok_or_else(|| ParseError("int: missing \"value\"".to_string()))?;
            Ok(LogicalValue::Int(v.to_string()))
        }
        "float" => {
            let width = obj.get("width").and_then(Value::as_str)
                .ok_or_else(|| ParseError("float: missing \"width\"".to_string()))?;
            let value = obj.get("value").and_then(Value::as_str)
                .ok_or_else(|| ParseError("float: missing \"value\"".to_string()))?;
            Ok(LogicalValue::Float { width: width.to_string(), value: value.to_string() })
        }
        "text" => {
            let v = obj.get("value").and_then(Value::as_str)
                .ok_or_else(|| ParseError("text: missing \"value\"".to_string()))?;
            Ok(LogicalValue::Text(v.to_string()))
        }
        "bytes" => {
            let v = obj.get("value").and_then(Value::as_str)
                .ok_or_else(|| ParseError("bytes: missing \"value\"".to_string()))?;
            Ok(LogicalValue::Bytes(v.to_string()))
        }
        "bool" => {
            let v = obj.get("value").and_then(Value::as_bool)
                .ok_or_else(|| ParseError("bool: missing \"value\"".to_string()))?;
            Ok(LogicalValue::Bool(v))
        }
        "null" => Ok(LogicalValue::Null),
        "array" => {
            let items = obj.get("items").and_then(Value::as_array)
                .ok_or_else(|| ParseError("array: missing \"items\"".to_string()))?;
            let parsed = items.iter().map(parse).collect::<Result<Vec<_>, _>>()?;
            Ok(LogicalValue::Array(parsed))
        }
        "map" => {
            let entries = obj.get("entries").and_then(Value::as_array)
                .ok_or_else(|| ParseError("map: missing \"entries\"".to_string()))?;
            let mut parsed = Vec::with_capacity(entries.len());
            for entry in entries {
                let pair = entry.as_array()
                    .ok_or_else(|| ParseError("map entry must be a 2-element array".to_string()))?;
                if pair.len() != 2 {
                    return Err(ParseError("map entry must have exactly 2 items".to_string()));
                }
                parsed.push((parse(&pair[0])?, parse(&pair[1])?));
            }
            Ok(LogicalValue::Map(parsed))
        }
        "tag" => {
            let tag = obj.get("tag").and_then(Value::as_u64)
                .ok_or_else(|| ParseError("tag: missing \"tag\" number".to_string()))?;
            let value = obj.get("value")
                .ok_or_else(|| ParseError("tag: missing \"value\"".to_string()))?;
            Ok(LogicalValue::Tag(tag, Box::new(parse(value)?)))
        }
        "bignum" => {
            let sign = obj.get("sign").and_then(Value::as_str)
                .ok_or_else(|| ParseError("bignum: missing \"sign\"".to_string()))?;
            let value = obj.get("value").and_then(Value::as_str)
                .ok_or_else(|| ParseError("bignum: missing \"value\"".to_string()))?;
            Ok(LogicalValue::Bignum { sign: sign.to_string(), value: value.to_string() })
        }
        other => Err(ParseError(format!("unknown logical-value type: {other:?}"))),
    }
}

pub fn to_json(value: &LogicalValue) -> Value {
    match value {
        LogicalValue::Int(s) => serde_json::json!({"type": "int", "value": s}),
        LogicalValue::Float { width, value } => {
            serde_json::json!({"type": "float", "width": width, "value": value})
        }
        LogicalValue::Text(s) => serde_json::json!({"type": "text", "value": s}),
        LogicalValue::Bytes(hex) => serde_json::json!({"type": "bytes", "value": hex}),
        LogicalValue::Bool(b) => serde_json::json!({"type": "bool", "value": b}),
        LogicalValue::Null => serde_json::json!({"type": "null"}),
        LogicalValue::Array(items) => {
            serde_json::json!({"type": "array", "items": items.iter().map(to_json).collect::<Vec<_>>()})
        }
        LogicalValue::Map(entries) => {
            let entries: Vec<Value> = entries.iter().map(|(k, v)| serde_json::json!([to_json(k), to_json(v)])).collect();
            serde_json::json!({"type": "map", "entries": entries})
        }
        LogicalValue::Tag(tag, inner) => {
            serde_json::json!({"type": "tag", "tag": tag, "value": to_json(inner)})
        }
        LogicalValue::Bignum { sign, value } => {
            serde_json::json!({"type": "bignum", "sign": sign, "value": value})
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn parses_int() {
        let v = parse(&json!({"type": "int", "value": "42"})).unwrap();
        assert_eq!(v, LogicalValue::Int("42".to_string()));
    }

    #[test]
    fn parses_float() {
        let v = parse(&json!({"type": "float", "width": "auto", "value": "2.5"})).unwrap();
        assert_eq!(v, LogicalValue::Float { width: "auto".to_string(), value: "2.5".to_string() });
    }

    #[test]
    fn parses_text() {
        let v = parse(&json!({"type": "text", "value": "café"})).unwrap();
        assert_eq!(v, LogicalValue::Text("café".to_string()));
    }

    #[test]
    fn parses_bytes() {
        let v = parse(&json!({"type": "bytes", "value": "deadbeef"})).unwrap();
        assert_eq!(v, LogicalValue::Bytes("deadbeef".to_string()));
    }

    #[test]
    fn parses_bool_and_null() {
        assert_eq!(parse(&json!({"type": "bool", "value": true})).unwrap(), LogicalValue::Bool(true));
        assert_eq!(parse(&json!({"type": "null"})).unwrap(), LogicalValue::Null);
    }

    #[test]
    fn parses_array() {
        let v = parse(&json!({"type": "array", "items": [
            {"type": "int", "value": "1"},
            {"type": "int", "value": "2"}
        ]})).unwrap();
        assert_eq!(v, LogicalValue::Array(vec![
            LogicalValue::Int("1".to_string()),
            LogicalValue::Int("2".to_string()),
        ]));
    }

    #[test]
    fn parses_map() {
        let v = parse(&json!({"type": "map", "entries": [
            [{"type": "text", "value": "a"}, {"type": "int", "value": "1"}]
        ]})).unwrap();
        assert_eq!(v, LogicalValue::Map(vec![
            (LogicalValue::Text("a".to_string()), LogicalValue::Int("1".to_string())),
        ]));
    }

    #[test]
    fn parses_tag() {
        let v = parse(&json!({"type": "tag", "tag": 100, "value": {"type": "int", "value": "5"}})).unwrap();
        assert_eq!(v, LogicalValue::Tag(100, Box::new(LogicalValue::Int("5".to_string()))));
    }

    #[test]
    fn parses_bignum() {
        let v = parse(&json!({"type": "bignum", "sign": "positive", "value": "18446744073709551616"})).unwrap();
        assert_eq!(v, LogicalValue::Bignum { sign: "positive".to_string(), value: "18446744073709551616".to_string() });
    }

    #[test]
    fn rejects_unknown_type() {
        assert!(parse(&json!({"type": "nonsense"})).is_err());
    }

    #[test]
    fn rejects_missing_type() {
        assert!(parse(&json!({})).is_err());
    }
}
